MetaMask/metamask-extension
# Fix EIP-1559 network compatibility check

## Explanation


This PR aims to fix a bug that occurs when a user switches the network and performs a transaction, triggering the check for EIP-1559 compatibility. The bug causes an incorrect determination of whether the network is compatible or not with EIP-1559. This PR addresses the issue by not assigning incorrectly the compatibility with EIP-1559 when middleware returns `null` getting the latest block. If the above happens from the perspective of the user it won't change anything and as the compatibility with EIP-1559 was not determined the first time due to being unable to retrieve the latest block, it will try to determine again, [here](https://github.com/MetaMask/core/blob/main/packages/network-controller/src/NetworkController.ts#L918).

Note: the extension is now using the NetworkController from core repo so the changes in the code from this PR were moved to [PR](https://github.com/MetaMask/core/pull/1457) keeping here only the new `@metamask/network-controller` version.

## Screenshots/Screencaps



### Before


Once the user switches the network even when the bug happens and the middleware resolves the latest block with `null` the network switch used to be concluded successfully with network status `available` but the `EIPS[1559]` in `networkDetails` would always return `false` even when the network supports the EIP-1559.

So the next time the compatibility is checked is when you perform a transaction using EIP-1559 and the check returns the `EIPS[1559]` in `networkDetails`  incorrectly determined (in this case) when the network switch happened, throwing an error mentioned in the bug.

![Screenshot from 2023-05-26 16-54-04](https://github.com/MetaMask/metamask-extension/assets/45455812/28cf0b31-0e75-412d-83fb-ee9b12d5b351)


### After
If middleware returns `null` getting the latest block from the perspective of the user, it won't change anything but the compatibility with EIP-1559 will not be set  and when the user performs a transaction as the compatibility with EIP-1559 was not determined the first time due to being unable to retrieve the latest block, it will try to determine it again, [here](https://github.com/MetaMask/core/blob/main/packages/network-controller/src/NetworkController.ts#L918)

[simulate-middleware-resolves-latest-block-null-on-switch-network.webm](https://github.com/MetaMask/core/assets/45455812/6c6dc9b2-f0e2-489f-aa0a-18504edc561f)


**Leaving the first proposal as a reference**
It retries 3 times until the middleware provides the latest block so the check can be performed. If after the retries the block is still null It throws and prevents the user to switch the network. It's the behaviour expected if the middleware/provider responds with an error.
[middleware-resolves-last-block-null](https://github.com/MetaMask/metamask-extension/assets/45455812/5d1627bf-2659-4655-b0da-a84ee73e5481)





## Manual Testing Steps


- Open MetaMask and switch to a network compatible with EIP-1559 (sepolia/goerli)
- The switch should occur as normal
- Go to [the test dapp](https://metamask.github.io/test-dapp/)
- Perform a transaction using the "Send EIP 1559 Transaction" button
- Transaction should be submitted as normal

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
